File: src/game/races/assignRace.ts
import { withRecalculatedVitals } from '../equipment/vitals'
import type { GameDatabase } from '../data/types'
import type { PlayerSave } from '../save/types'
import { grantRaceStartingItems, raceById } from './races'

export type AssignRaceResult =
  | { ok: true; save: PlayerSave; grantedStarterKit: boolean }
  | { ok: false; reason: string }

/**
 * Set the player's race. When moving from no race → a race (first pick),
 * grants that race's starting item kit. Later race changes (menu test /
 * future quest) only swap raceId and refresh vitals — kits are not re-granted.
 */
export function assignRace(
  db: GameDatabase,
  save: PlayerSave,
  raceId: string,
): AssignRaceResult {
  const race = raceById(db, raceId)
  if (!race) return { ok: false, reason: 'Unknown race.' }

  const firstSelection = save.raceId == null
  let next: PlayerSave = { ...save, raceId }
  if (firstSelection) {
    next = grantRaceStartingItems(db, next, raceId)
  }
  next = withRecalculatedVitals(db, next)
  return { ok: true, save: next, grantedStarterKit: firstSelection }
}

File: packages/ik_rules/lib/src/races/assign_race.dart
import 'package:ik_content/ik_content.dart';

import '../equipment/vitals.dart';
import '../save/generated/save_models.dart';
import 'races.dart';

class AssignRaceResult {
  const AssignRaceResult.ok(this.save, {required this.grantedStarterKit}) : reason = null;

  const AssignRaceResult.failed(this.reason) : save = null, grantedStarterKit = false;

  final PlayerSave? save;
  final bool grantedStarterKit;
  final String? reason;

  bool get ok => reason == null;
}

/// Sets the player's race.
///
/// When moving from no race → a race (first pick), grants that race's starting
/// item kit. Later race changes (menu test / future quest) only swap raceId and
/// refresh vitals — kits are not re-granted.
AssignRaceResult assignRace(GameDatabase db, PlayerSave save, String raceId) {
  if (raceById(db, raceId) == null) {
    return const AssignRaceResult.failed('Unknown race.');
  }

  final firstSelection = save.raceId == null;
  var next = save.copyWith(raceId: raceId);
  if (firstSelection) {
    next = grantRaceStartingItems(db, next, raceId);
  }
  return AssignRaceResult.ok(withRecalculatedVitals(db, next), grantedStarterKit: firstSelection);
}
